### PR TITLE
Enable LLaVA-Next in VLM pipeline

### PR DESCRIPTION
**Pull request by yatarkan**
This PR adds:
- [x] Basic support of LLaVA-Next models (`llava_v1_6_mistral`, `llava_v1_6_vicuna_7b` and `llama3_llava_next_8b`)
- [x] Add C++ sample test with LLaVA-Next model

Common TODOs for LLaVa models in VLM pipeline:

- [ ] Handle disabled chat mode (llava, llava next)
- [ ] Refactor clip and vision encoder (refine, restructure, remove unused functions/class members)
- [ ] Process image patches as a single batch
- [ ] Support multiple image input (llava, llavanext)

Current Limitations:
- LLaVA and LLaVA-Next models support only single image input

Ticket: CVS-153333

### CODE REVIEW

File: .github/workflows/causal_lm_cpp.yml [hunk 1]
```diff
@@ -723,6 +723,19 @@ jobs:
           && ./build/samples/cpp/visual_language_chat/visual_language_chat ./llava_1_5_7b_ov/ monalisa.jpg
           <<< $'Who drew this painting?\nWhen did the painter live?'
         timeout-minutes: 4
+      - name: Download and convert LLaVa-Next model and an image
+        run: |
+          source ./ov/setupvars.sh
+          python -m pip install ./thirdparty/openvino_tokenizers/[transformers] --pre --extra-index-url https://storage.openvinotoolkit.org/simple/wheels/nightly
+          python -m pip install --upgrade-strategy eager -r ./samples/requirements.txt --pre --extra-index-url https://storage.openvinotoolkit.org/simple/wheels/nightly
```
Comment (ilya-lavrenov) [inline .github/workflows/causal_lm_cpp.yml:730]: dependencies are already installed above (twice)
maybe we can split install dependencies step and then download models minicpm, LLaVA 1.5, LLaVa next

File: src/cpp/src/visual_language/vision_encoder.cpp [hunk 2]
```diff
@@ -417,23 +410,81 @@ ov::Tensor preprocess_image_llava(const ov::Tensor& image, const ProcessorConfig
...
+    std::pair<int, int> size{config.size_shortest_edge, config.size_shortest_edge};
+    auto patch_size = config.crop_size_height;
+    auto image_patches = get_image_patches(input_image, config.image_grid_pinpoints, size, patch_size);
+
+    // Preprocess image patches
+    std::vector<clip_image_f32> processed_patches;
+    processed_patches.reserve(image_patches.size());
+
+    for (const auto& patch : image_patches) {
+        processed_patches.push_back(preprocess_clip_image_llava(patch, config));
```
Comment (ilya-lavrenov) [inline src/cpp/src/visual_language/vision_encoder.cpp:454]: is it possible to process image patches as a single batch ? just as a future optimization

File: src/cpp/src/visual_language/pipeline.cpp [hunk 3]
```diff
@@ -556,9 +818,51 @@ class ov::genai::VLMPipeline::VLMPipelineImpl {
...
+        } catch (const std::exception& error) {
+            // TODO Consider using template syntax instead of concatenating
+            new_templated_chat_history = m_tokenizer.apply_chat_template(m_history, add_generation_prompt, "USER: " + content + " ASSISTANT:");
+        }
+         
+        ov::Tensor input_ids = m_tokenizer.encode(new_templated_chat_history).input_ids;
+        if (images.empty()) {
+            return process_prompt(m_embedding, input_ids, m_vlm_config.scale_emb);
+        } else {
+            OPENVINO_ASSERT(1 == images.size(), "Only a single image allowed");
```
Comment (ilya-lavrenov) [inline src/cpp/src/visual_language/pipeline.cpp:847]: is it a current limitation of limitation of the model itself?

File: src/cpp/src/visual_language/pipeline.cpp [hunk 4]
```diff
@@ -544,7 +805,8 @@ class ov::genai::VLMPipeline::VLMPipelineImpl {
     }
 
     ov::Tensor get_inputs_embeds_llava(const std::string& prompt, const std::vector<ov::Tensor>& images) {
-        std::string image_token = "<image>"; // TODO Consider getting from vlm_config or json
+        std::string image_token = m_vlm_config.im_start;
+        // TODO Enable and reuse chat mode
```
Comment (ilya-lavrenov) [inline src/cpp/src/visual_language/pipeline.cpp:809]: should we create a list of TODOs in PR description as follow up for future PRs?

File: .github/workflows/causal_lm_cpp.yml (hunk 1 above)
Comment (Wovchena) [inline .github/workflows/causal_lm_cpp.yml:730]:
```suggestion
```

File: src/cpp/src/visual_language/clip.cpp [hunk 5]
```diff
@@ -207,7 +207,8 @@ static void resize_and_pad_image(const clip_image_u8& image, clip_image_u8 &imag
  * @param possible_resolutions A list of possible resolutions in the format [(width1, height1), (width2, height2), ...].
  * @return The best fit resolution in the format (width, height).
  */
-static std::pair<int, int> select_best_resolution(const std::pair<int, int> & original_size, const std::vector<std::pair<int, int>> & possible_resolutions) {
+std::pair<int, int> select_best_resolution(const std::pair<int, int> & original_size, const std::vector<std::pair<int, int>> & possible_resolutions) {
+    // TODO Consider changing original_size and return value to (height, width) format
```
Comment (Wovchena) [inline src/cpp/src/visual_language/clip.cpp:211]: What does the comment mean? the function already returns a height width pair.

File: src/cpp/src/visual_language/pipeline.cpp [hunk 6]
```diff
@@ -556,9 +818,51 @@ class ov::genai::VLMPipeline::VLMPipelineImpl {
...
+            ov::Tensor encoded_image_token = m_tokenizer.encode(image_token, ov::genai::add_special_tokens(false)).input_ids;
+            int64_t image_token_id = encoded_image_token.data<int64_t>()[encoded_image_token.get_size() - 1];
+
+            return merge_text_and_image_embeddings_llava(input_ids, text_embeds, image_embeds, image_token_id);
+        }
+    }
+
+    ov::Tensor get_inputs_embeds_llava_next(const std::string& prompt, const std::vector<ov::Tensor>& images) {
+        std::string image_token = m_vlm_config.im_start;
+        // TODO Enable and reuse chat mode
```
Comment (ilya-lavrenov) [inline src/cpp/src/visual_language/pipeline.cpp:830]: it it valid comment? looks like chat mode is here

File: src/cpp/src/visual_language/vision_encoder.cpp [hunk 7]
```diff
@@ -363,28 +363,21 @@ ProcessorConfig from_any_map(
     return extracted_config;
 }
 
-
-ov::Tensor preprocess_image_llava(const ov::Tensor& image, const ProcessorConfig& config) {
+// TODO Consider moving to clip.cpp or separate dedicated file
```
Comment (Wovchena) [inline src/cpp/src/visual_language/vision_encoder.cpp:366]: That isn't something that is hard to do. Either do it now or remove the comment. Same below

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: src/cpp/src/visual_language/clip.cpp (hunk 5 above)
Comment (yatarkan) [reply src/cpp/src/visual_language/clip.cpp:211]: Now the function returns (width, height) pair, maybe it makes sense to change to inverted (height, width) pair for consistency with other processing logic

File: src/cpp/src/visual_language/vision_encoder.cpp (hunk 7 above)
Comment (yatarkan) [reply src/cpp/src/visual_language/vision_encoder.cpp:366]: As we were planning to refactor, removing for now

File: src/cpp/src/visual_language/pipeline.cpp (hunk 3 above)
Comment (yatarkan) [reply src/cpp/src/visual_language/pipeline.cpp:847]: This is current limitation

File: src/cpp/src/visual_language/pipeline.cpp [hunk 8]
```diff
@@ -556,9 +818,51 @@ class ov::genai::VLMPipeline::VLMPipelineImpl {
...
+        std::string content = images.empty() ? prompt : image_token + "\n" + prompt;
+        ChatHistory m_history;
```
Comment (ilya-lavrenov) [inline src/cpp/src/visual_language/pipeline.cpp:832]:
```suggestion
```
`m_history` is already part of class

Please, also handle when chat mode is disabled

File: .github/workflows/causal_lm_cpp.yml (hunk 1 above)
Comment (yatarkan) [reply .github/workflows/causal_lm_cpp.yml:730]: Moved to separate step

Comment (yatarkan) [reply .github/workflows/causal_lm_cpp.yml:730]: Fixed

**Comment by Wovchena** [comment]
In order to merge it faster, let's merge now. Add fixes later if any.

File: src/cpp/src/visual_language/vision_encoder.cpp (hunk 2 above)
Comment (yatarkan) [reply src/cpp/src/visual_language/vision_encoder.cpp:452]: Added to PR TODOS

File: src/cpp/src/visual_language/pipeline.cpp (hunk 3 above)
Comment (yatarkan) [reply src/cpp/src/visual_language/pipeline.cpp:844]: Added to PR TODOs

File: src/cpp/src/visual_language/pipeline.cpp (hunk 8 above)
Comment (yatarkan) [reply src/cpp/src/visual_language/pipeline.cpp]: Added to PR TODOs

**Review by ilya-lavrenov — APPROVED** [review]

File: .github/workflows/causal_lm_cpp.yml [hunk 9]
```diff
@@ -698,16 +698,18 @@ jobs:
...
-      - name: Download and convert MiniCPM-V-2_6 model and an image
+      - name: Install dependencies
         run: |
           source ./ov/setupvars.sh
           python -m pip install ./thirdparty/openvino_tokenizers/[transformers] --pre --extra-index-url https://storage.openvinotoolkit.org/simple/wheels/nightly
           python -m pip install --upgrade-strategy eager -r ./samples/requirements.txt --pre --extra-index-url https://storage.openvinotoolkit.org/simple/wheels/nightly
+      - name: Download and convert MiniCPM-V-2_6 model and an image
+        run: |
           python -m pip install git+https://github.com/eaidova/optimum-intel.git@ea/minicpmv
           python -m pip install -U "optimum<1.23" --no-dependencies
```
Comment (Wovchena) [inline .github/workflows/causal_lm_cpp.yml:709]:
```suggestion
          python -m pip install -U "optimum<1.23" --no-dependencies
          source ./ov/setupvars.sh
```

**Pull request merged**